data-prepper-api/src/main/java/org/opensearch/dataprepper/model/event/Event.java

```diff
@@ -143,11 +143,12 @@ public interface Event extends Serializable {
      * of a Data Prepper expression
      * @param format input format
      * @param expressionEvaluator - The expression evaluator that will support formatting from Data Prepper expressions
+     * @param replacementForFailures - The String to use as a replacement for when keys in Events can't be found
      * @return returns a string with no formatted parts, returns null if no value is found
      * @throws RuntimeException if the input string is not properly formatted
      * @since 2.1
      */
-    String formatString(String format, ExpressionEvaluator expressionEvaluator);
+    String formatString(final String format, final ExpressionEvaluator expressionEvaluator, final String replacementForFailures);
```

Comment:
You can override this to avoid the changes that have the explicit `null`.

```
String formatString(final String format, final ExpressionEvaluator expressionEvaluator, final String replacementForFailures);

default String formatString(final String format, final ExpressionEvaluator expressionEvaluator) {
  return formatString(format, expressionEvaluator, null);
}
```